File: src/v2/moves.py
# -*- coding: utf-8 -*-
"""v2 move encoding: AlphaZero-style 8x8x73 = 4672 move space.

For each of the 64 from-squares, 73 possible "move types":
  - 56 sliding moves: 8 directions x 7 distances (1..7)
    direction order: N=0, NE=1, E=2, SE=3, S=4, SW=5, W=6, NW=7
    queen-promotions are encoded as the regular sliding move to last rank
  - 8 knight moves: 8 (dr, df) offsets in a fixed order
  - 9 underpromotions: 3 directions (left-diagonal, forward, right-diagonal)
    x 3 promotion pieces (knight, bishop, rook). Queen-promos use the
    regular sliding move (above), not this block.

Flat index convention: idx = from_square * 73 + move_type
where from_square is the python-chess square index (a1=0..h8=63).

So the model's output tensor (B, 73, 8, 8) — channel = move_type,
spatial = from_square (h*8 + w == from_square in python-chess).
When flattened it gives flat[from_sq * 73 + move_type].
"""
import chess
import numpy as np

# Sliding directions: (dh, dw) per direction index 0..7
# h = rank (0..7), w = file (0..7); +h is "up the board for white"
_SLIDE_DIRS = [
    (+1,  0),  # 0: N
    (+1, +1),  # 1: NE
    ( 0, +1),  # 2: E
    (-1, +1),  # 3: SE
    (-1,  0),  # 4: S
    (-1, -1),  # 5: SW
    ( 0, -1),  # 6: W
    (+1, -1),  # 7: NW
]
NUM_SLIDE_TYPES = 8 * 7  # 56

# Knight offsets in a fixed canonical order
_KNIGHT_DIRS = [
    (+2, +1), (+2, -1), (-2, +1), (-2, -1),
    (+1, +2), (+1, -2), (-1, +2), (-1, -2),
]
NUM_KNIGHT_TYPES = 8

# Underpromotion: (dh_relative, dw_relative, promotion_piece)
# For white promoting to rank 8: dh = +1; dw in {-1, 0, +1} (capture-left,
# forward, capture-right). Promotion piece is N (knight), B (bishop), or R (rook).
# Queen-promotions use the regular sliding move (sliding N, distance 1).
# For black, dh = -1, but the move-type is still indexed by (relative direction)
# because the move-type space is fixed at 73 regardless of color.
_UNDERPROM_DIRS = [-1, 0, +1]   # df offset (left, forward, right)
_UNDERPROM_PIECES = [chess.KNIGHT, chess.BISHOP, chess.ROOK]
NUM_UNDERPROM_TYPES = 9

NUM_MOVE_TYPES = NUM_SLIDE_TYPES + NUM_KNIGHT_TYPES + NUM_UNDERPROM_TYPES  # 73
NUM_MOVES = 64 * NUM_MOVE_TYPES  # 4672


# ---- Build lookup tables once (encoder + decoder) ----

def _build_encode_table():
    """Returns dict: (from_sq, to_sq, promotion_or_None) -> move_type_index"""
    table = {}

    # Sliding moves
    for from_sq in range(64):
        fh, fw = chess.square_rank(from_sq), chess.square_file(from_sq)
        for d_idx, (dh, dw) in enumerate(_SLIDE_DIRS):
            for dist in range(1, 8):
                th = fh + dh * dist
                tw = fw + dw * dist
                if 0 <= th < 8 and 0 <= tw < 8:
                    to_sq = th * 8 + tw
                    move_type = d_idx * 7 + (dist - 1)
                    table[(from_sq, to_sq, None)] = move_type
                    # Queen-promotion uses the sliding move encoding too
                    # (only matters when from on rank 7 white / rank 1 black)
                    table[(from_sq, to_sq, chess.QUEEN)] = move_type

    # Knight moves
    for from_sq in range(64):
        fh, fw = chess.square_rank(from_sq), chess.square_file(from_sq)
        for k_idx, (dh, dw) in enumerate(_KNIGHT_DIRS):
            th = fh + dh
            tw = fw + dw
            if 0 <= th < 8 and 0 <= tw < 8:
                to_sq = th * 8 + tw
                move_type = NUM_SLIDE_TYPES + k_idx
                table[(from_sq, to_sq, None)] = move_type

    # Underpromotions (N, B, R only — Q uses sliding)
    for from_sq in range(64):
        fh, fw = chess.square_rank(from_sq), chess.square_file(from_sq)
        # Determine promotion direction by source rank: white promotes from
        # rank 7 (move +1), black promotes from rank 1 (move -1)
        for dh in (+1, -1):
            th = fh + dh
            if th != 0 and th != 7:
                continue  # underpromotion only into last-rank
            for df_idx, df in enumerate(_UNDERPROM_DIRS):
                tw = fw + df
                if not (0 <= tw < 8):
                    continue
                to_sq = th * 8 + tw
                for p_idx, piece in enumerate(_UNDERPROM_PIECES):
                    move_type = (NUM_SLIDE_TYPES + NUM_KNIGHT_TYPES
                                 + df_idx * 3 + p_idx)
                    # Same (from, to, piece) keyed by promotion piece
                    table[(from_sq, to_sq, piece)] = move_type

    return table


_ENCODE_TABLE = _build_encode_table()


def _build_decode_table():
    """Returns array of shape (64, 73, 3) -> (from_sq, to_sq, promotion_or_-1)
    where promotion is -1 if not a promotion, else the chess piece type int.
    Used to convert (from_sq, move_type) back to a chess.Move quickly.
    """
    table = np.full((64, NUM_MOVE_TYPES, 3), -1, dtype=np.int8)
    for (from_sq, to_sq, prom), mt in _ENCODE_TABLE.items():
        # Sliding & queen-promotion share a slot; the inverse for that slot
        # records the no-promotion-needed sliding move. Promotion is set
        # separately when the from-rank is pre-promotion.
        # For underpromotions specifically the slot is unambiguous.
        # We'll just take the last writer; underpromotion entries are written
        # after sliding so they overwrite cleanly for those move_type slots
        # but those move_types don't overlap with sliding anyway.
        table[from_sq, mt, 0] = from_sq
        table[from_sq, mt, 1] = to_sq
        table[from_sq, mt, 2] = -1 if prom is None or prom == chess.QUEEN else prom
    return table


_DECODE_TABLE = _build_decode_table()


def encode_move(move: chess.Move) -> int:
    """Convert a chess.Move to its flat index in [0, 4672).

    Returns -1 if the move can't be encoded (shouldn't happen for legal moves
    but defensive — e.g., null moves).
    """
    if move is None or move.from_square is None or move.to_square is None:
        return -1
    key = (move.from_square, move.to_square, move.promotion)
    mt = _ENCODE_TABLE.get(key, -1)
    if mt < 0:
        return -1
    return move.from_square * NUM_MOVE_TYPES + mt


def decode_move(flat_idx: int, board: chess.Board) -> chess.Move:
    """Convert a flat move index back to a chess.Move on the given board.

    The board is needed to:
      - Disambiguate queen-promotion vs regular sliding move for pawns
        reaching the last rank (the move_type slot is the same).
      - Always returns a chess.Move object; legality is the caller's problem.
    """
    from_sq = flat_idx // NUM_MOVE_TYPES
    mt = flat_idx % NUM_MOVE_TYPES
    to_sq = int(_DECODE_TABLE[from_sq, mt, 1])
    prom = int(_DECODE_TABLE[from_sq, mt, 2])
    if to_sq < 0:
        # Unmapped slot (shouldn't be hit for indexes from a softmax)
        return chess.Move.null()

    promotion = None if prom < 0 else prom

    # If this is a sliding move and the moving piece is a pawn reaching the
    # last rank, it MUST be a promotion. Default to queen (the underpromotion
    # paths use distinct move_type slots so we won't be here for those).
    if promotion is None:
        piece = board.piece_type_at(from_sq)
        if piece == chess.PAWN:
            to_rank = chess.square_rank(to_sq)
            if to_rank == 0 or to_rank == 7:
                promotion = chess.QUEEN

    return chess.Move(from_sq, to_sq, promotion=promotion)


def legal_mask(board: chess.Board) -> np.ndarray:
    """Return a (4672,) bool array marking which flat-indices are legal moves.

    Iterates board.legal_moves and encodes each. Used by the policy head's
    legal-move masking step (analogous to v1, but over the larger move space).
    """
    mask = np.zeros(NUM_MOVES, dtype=bool)
    for mv in board.legal_moves:
        idx = encode_move(mv)
        if idx >= 0:
            mask[idx] = True
    return mask

File: viz/src/core/engine/moves.ts
// Move encoding — faithful port of src/v2/moves.py (AlphaZero-style 8×8×73).
//
// Flat index = from_square*73 + move_type, from_square python-chess (a1=0..h8=63).
// 73 move types: 56 sliding (8 dir × 7 dist) + 8 knight + 9 underpromotion.
// Queen-promotions are encoded as the regular sliding move (NOT in the
// underpromotion block).
//
// rotateSquare(sq) = sq ^ 56 == chess.square_mirror == VERTICAL mirror (flip
// rank, keep file). Used to un-rotate model moves when black was to move. (NOT
// 63-sq, which would flip both axes — that is wrong; the engine is authoritative.)

import { PIECE_TYPE_INT, type BoardState } from './boardState.ts';

export const NUM_MOVE_TYPES = 73;
export const NUM_MOVES = 64 * NUM_MOVE_TYPES; // 4672

// (dh,dw) per direction 0..7: N,NE,E,SE,S,SW,W,NW. h=rank, w=file.
const SLIDE_DIRS: Array<[number, number]> = [
  [+1, 0],
  [+1, +1],
  [0, +1],
  [-1, +1],
  [-1, 0],
  [-1, -1],
  [0, -1],
  [+1, -1],
];
const NUM_SLIDE_TYPES = 56;

const KNIGHT_DIRS: Array<[number, number]> = [
  [+2, +1],
  [+2, -1],
  [-2, +1],
  [-2, -1],
  [+1, +2],
  [+1, -2],
  [-1, +2],
  [-1, -2],
];
const NUM_KNIGHT_TYPES = 8;

const UNDERPROM_DIRS = [-1, 0, +1]; // df: left, forward, right
// python-chess piece ints: KNIGHT=2, BISHOP=3, ROOK=4
const UNDERPROM_PIECES = [2, 3, 4];

const QUEEN = 5; // python-chess QUEEN

export interface Move {
  from: number;
  to: number;
  /** python-chess promotion piece int (2=N,3=B,4=R,5=Q) or null/undefined. */
  promotion?: number | null;
}

// (from,to,prom) -> move_type. prom key: 0 = none, else piece int.
const ENCODE = new Map<number, number>();
// decode[from*73+mt] = {to, prom(-1 if none/queen-slide)}
const DECODE_TO = new Int32Array(NUM_MOVES).fill(-1);
const DECODE_PROM = new Int32Array(NUM_MOVES).fill(-1);

function encKey(from: number, to: number, prom: number): number {
  // from,to in [0,64), prom in [0,6]
  return (from * 64 + to) * 8 + prom;
}

function rankOf(sq: number): number {
  return sq >> 3;
}
function fileOf(sq: number): number {
  return sq & 7;
}

(function buildTables() {
  // Sliding moves (+ queen-promotion sharing the slot)
  for (let from = 0; from < 64; from++) {
    const fh = rankOf(from);
    const fw = fileOf(from);
    for (let d = 0; d < SLIDE_DIRS.length; d++) {
      const [dh, dw] = SLIDE_DIRS[d];
      for (let dist = 1; dist < 8; dist++) {
        const th = fh + dh * dist;
        const tw = fw + dw * dist;
        if (th < 0 || th >= 8 || tw < 0 || tw >= 8) continue;
        const to = th * 8 + tw;
        const mt = d * 7 + (dist - 1);
        ENCODE.set(encKey(from, to, 0), mt);
        ENCODE.set(encKey(from, to, QUEEN), mt);
        DECODE_TO[from * 73 + mt] = to;
        DECODE_PROM[from * 73 + mt] = -1;
      }
    }
  }
  // Knight moves
  for (let from = 0; from < 64; from++) {
    const fh = rankOf(from);
    const fw = fileOf(from);
    for (let kI = 0; kI < KNIGHT_DIRS.length; kI++) {
      const [dh, dw] = KNIGHT_DIRS[kI];
      const th = fh + dh;
      const tw = fw + dw;
      if (th < 0 || th >= 8 || tw < 0 || tw >= 8) continue;
      const to = th * 8 + tw;
      const mt = NUM_SLIDE_TYPES + kI;
      ENCODE.set(encKey(from, to, 0), mt);
      DECODE_TO[from * 73 + mt] = to;
      DECODE_PROM[from * 73 + mt] = -1;
    }
  }
  // Underpromotions (N,B,R only)
  for (let from = 0; from < 64; from++) {
    const fh = rankOf(from);
    const fw = fileOf(from);
    for (const dh of [+1, -1]) {
      const th = fh + dh;
      if (th !== 0 && th !== 7) continue;
      for (let dfI = 0; dfI < UNDERPROM_DIRS.length; dfI++) {
        const tw = fw + UNDERPROM_DIRS[dfI];
        if (tw < 0 || tw >= 8) continue;
        const to = th * 8 + tw;
        for (let pI = 0; pI < UNDERPROM_PIECES.length; pI++) {
          const piece = UNDERPROM_PIECES[pI];
          const mt = NUM_SLIDE_TYPES + NUM_KNIGHT_TYPES + dfI * 3 + pI;
          ENCODE.set(encKey(from, to, piece), mt);
          DECODE_TO[from * 73 + mt] = to;
          DECODE_PROM[from * 73 + mt] = piece;
        }
      }
    }
  }
})();

/** Encode a move to its flat index in [0,4672), or -1 if unencodable. */
export function encodeMove(move: Move): number {
  if (move.from == null || move.to == null) return -1;
  const prom = move.promotion ?? 0;
  const mt = ENCODE.get(encKey(move.from, move.to, prom));
  if (mt === undefined) return -1;
  return move.from * NUM_MOVE_TYPES + mt;
}

/**
 * Decode a flat index back to a Move on the given board. The board disambiguates
 * queen-promotion vs a plain sliding move for a pawn reaching the last rank.
 */
export function decodeMove(flatIdx: number, board: BoardState): Move {
  const from = Math.floor(flatIdx / NUM_MOVE_TYPES);
  const mt = flatIdx % NUM_MOVE_TYPES;
  const to = DECODE_TO[from * 73 + mt];
  if (to < 0) return { from, to: from, promotion: null }; // unmapped slot
  let promotion: number | null = DECODE_PROM[from * 73 + mt] < 0 ? null : DECODE_PROM[from * 73 + mt];
  if (promotion === null) {
    const pt = board.pieceTypeAt(from);
    if (pt === 'p') {
      const toRank = rankOf(to);
      if (toRank === 0 || toRank === 7) promotion = QUEEN;
    }
  }
  return { from, to, promotion };
}

/** Build a (4672,) legal mask from a list of legal moves (encoding each). */
export function legalMask(legalMoves: Move[]): Uint8Array {
  const mask = new Uint8Array(NUM_MOVES);
  for (const mv of legalMoves) {
    const idx = encodeMove(mv);
    if (idx >= 0) mask[idx] = 1;
  }
  return mask;
}

/**
 * Softmax over ONLY the legal flat indices → Map(flatIdx → probability), summing
 * to 1. Mathematically identical to a full-domain softmax then mask+renormalize
 * over the legal set (the engine's policyProbs), but computed straight from the
 * legal indices so it can be derived from just the logits + a legal-move list.
 */
export function legalPolicySoftmax(logits: ArrayLike<number>, legalIndices: number[]): Map<number, number> {
  const out = new Map<number, number>();
  if (legalIndices.length === 0) return out;
  let max = -Infinity;
  for (const i of legalIndices) if (logits[i] > max) max = logits[i];
  const weights: number[] = [];
  let total = 0;
  for (const i of legalIndices) {
    const e = Math.exp(logits[i] - max);
    weights.push(e);
    total += e;
  }
  total = total || 1;
  legalIndices.forEach((i, k) => out.set(i, weights[k] / total));
  return out;
}

/**
 * P(the model moves the piece on each FROM-square) — for each from-square, the
 * SUM of the legal-masked, renormalized policy probability over the 73 move-types
 * originating from that square. Returns a 64-array indexed by from-square
 * (a1=0 .. h8=63), in whatever frame the supplied indices are in (engine frame
 * for the policy head). SUM, not average: a square's total move probability.
 */
export function pieceToMoveProbs(probsByIndex: Map<number, number>): Float32Array {
  const out = new Float32Array(64);
  for (const [idx, p] of probsByIndex) out[Math.floor(idx / NUM_MOVE_TYPES)] += p;
  return out;
}

/** Vertical-mirror a square (flip rank, keep file): chess.square_mirror. */
export function rotateSquare(sq: number): number {
  return sq ^ 56;
}

export { PIECE_TYPE_INT };
